Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,702 @@
+    ref->local = udpdata->source_port;
+    ref->remote = udpdata->destination_port;
+    ref->type = 17;
+    memcpy(&ref->bba_mac, &m_eth_ref->mBbaMem[BBA_NAFR_PAR0], Common::MAC_ADDRESS_SIZE);
+    ref->my_mac = fake_mac;
+    ref->from.sin_addr.s_addr = *(u32*)&ipdata->destination_addr;
+    ref->from.sin_port = udpdata->destination_port;
+    ref->to.sin_addr.s_addr = *(u32*)&ipdata->source_addr;
+    ref->to.sin_port = udpdata->source_port;
+    ref->udp_socket.setBlocking(false);
+    if (ref->udp_socket.bind(htons(udpdata->source_port)) != sf::Socket::Done)
+    {
+      if (ref->udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
+      {
+        ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+        return;
+      }
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 1900)
+  {
+    InitUDPPort(26512);                                 // MK DD and 1080
+    InitUDPPort(26502);                                 // Air Ride
+    if (*(u32*)ipdata->destination_addr == 0xFAFFFFEF)  // force real broadcast
+      *(u32*)ipdata->destination_addr = 0xFFFFFFFF;     // Multi cast cannot be read
+    if (udpdata->length > 150)
+    {
+      // Quick hack to unlock the connection, throw it back at him
+      Common::EthernetHeader* hwpart = (Common::EthernetHeader*)&m_in_frame[0];
+      Common::IPv4Header* ippart = (Common::IPv4Header*)&m_in_frame[14];
+      memcpy(&m_in_frame[0], hwdata, htons(ipdata->total_len) + 14);
+      hwpart->destination = hwdata->source;
+      hwpart->source = hwdata->destination;
+      *(u32*)ippart->destination_addr = *(u32*)ipdata->source_addr;
+      *(u32*)ippart->source_addr = *(u32*)ipdata->destination_addr;
+      WriteToQueue(&m_in_frame[0], htons(ipdata->total_len) + 14);
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 53)
+  {
+    target = sf::IpAddress(m_dns_ip.c_str());  // dns server ip
+  }
+  else
+  {
+    target = sf::IpAddress(Common::swap32(*(u32*)ipdata->destination_addr));
+  }
+  ref->udp_socket.send(data, Common::swap16(udpdata->length) - 8, target,
+                       Common::swap16(udpdata->destination_port));
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::SendFrame(const u8* frame, u32 size)
+{
+  Common::EthernetHeader* hwdata;
+  Common::IPv4Header* ipdata;
+  Common::UDPHeader* udpdata;
+  Common::TCPHeader* tcpdata;
+
+  int offset = 0;
+  std::lock_guard<std::mutex> lock(mtx);
+  memmove(&m_out_frame[0], frame, size);
+  INFO_LOG_FMT(SP1, "Sending packet {:x}\n{}", size, ArrayToString(&m_out_frame[0], size, 16));
+  // handle the packet data
+  hwdata = (Common::EthernetHeader*)&m_out_frame[0];
+  if (hwdata->ethertype == 0x08)  // IPV4
+  {
+    // IP sub
+    ipdata = (Common::IPv4Header*)&m_out_frame[14];
+    switch (ipdata->protocol)
+    {
+    case IPPROTO_UDP:
+      udpdata = (Common::UDPHeader*)&m_out_frame[14 + ((ipdata->version_ihl & 0xf) * 4)];
+      offset = 14 + ((ipdata->version_ihl & 0xf) * 4) + 8;
```

Comment:
Unnecessary parenthesis for the multiplication. The offset can be initialised above like so:
```c++
offset = 14 + (ipdata->version_ihl & 0xf) * 4;
udpdata = (Common::UDPHeader*)&m_out_frame[offset];
```
And incremented by 8 here:
```c++
offset += Common::UDPHeader::SIZE;
```